File: problems/033-toolchain-mirror-network/solutions/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

type E struct {
	u, v int
	w    uint64
}

func main() {
	in := bufio.NewReader(os.Stdin)
	var n, m int
	fmt.Fscan(in, &n, &m)
	e := make([]E, m)
	for i := range e {
		fmt.Fscan(in, &e[i].u, &e[i].v, &e[i].w)
		e[i].u--
		e[i].v--
	}
	sort.Slice(e, func(i, j int) bool { return e[i].w < e[j].w })
	p := make([]int, n)
	sz := make([]int, n)
	for i := range p {
		p[i] = i
		sz[i] = 1
	}
	find := func(x int) int { return 0 }
	find = func(x int) int {
		r := x
		for p[r] != r {
			r = p[r]
		}
		for p[x] != x {
			y := p[x]
			p[x] = r
			x = y
		}
		return r
	}
	var cost uint64
	take := 0
	for _, x := range e {
		u, v := find(x.u), find(x.v)
		if u == v {
			continue
		}
		if sz[u] < sz[v] {
			u, v = v, u
		}
		p[v] = u
		sz[u] += sz[v]
		cost += x.w
		take++
		if take == n-1 {
			break
		}
	}
	if take == n-1 {
		fmt.Println("COST", cost)
	} else {
		fmt.Println("IMPOSSIBLE")
	}
}

File: problems/033-toolchain-mirror-network/solutions/python/main.py
import sys

a = list(map(int, sys.stdin.read().split()))
n, m = a[:2]
e = [(a[i + 2], a[i] - 1, a[i + 1] - 1) for i in range(2, len(a), 3)]
e.sort()
p = list(range(n))
sz = [1] * n


def f(x):
    while p[x] != x:
        p[x] = p[p[x]]
        x = p[x]
    return x


cost = taken = 0
for w, u, v in e:
    u, v = f(u), f(v)
    if u == v:
        continue
    if sz[u] < sz[v]:
        u, v = v, u
    p[v] = u
    sz[u] += sz[v]
    cost += w
    taken += 1
    if taken == n - 1:
        break
print(f"COST {cost}" if taken == n - 1 else "IMPOSSIBLE")
